### DISCUSSION TITLE
How to override outdated "zls" binary in mason/packages/zls ?

### DISCUSSION THREAD

**Discussion by rseichter** [mason-org/mason.nvim#1283] [Q&A]
The `zls` binary currently bundled in mason/packages/zls is outdated (version 0.10.0-dev.364+7ef2244 at the time of opening this issue). I built a frsh `zls` binary from its source code, and I wonder what the correct mechanism to override the outdated binary is? Creating a softlink to `~/bin/zls` does not seem ideal. I use [LazyVim](https://www.lazyvim.org) and the following LUA initialisation code:
```lua
local lspconfig = require('lspconfig')
local on_attach = function(_, bufnr)
    vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
    require('completion').on_attach()
end
local servers = { 'zls' }
for _, lsp in ipairs(servers) do
    lspconfig[lsp].setup {
        on_attach = on_attach,
    }
end
return {}
```
Do you have any recommendations? Thank you.

**Comment by williamboman** [comment]
Hello! Modifying any of the files or directories that Mason creates is not recommended. If you want to install a version that cannot be installed via Mason, simply install it elsewhere on your system. Some common locations to put system-local installations are:

/opt
/usr/local
~/.local

**Comment by rseichter** [comment]
Thanks for your response. Is there a way to configure mason.nvim to use a specific version of `zls`, e.g. by specifying the full path to the binary I have built? Simply placing it in $PATH does not work, the plugin still loads the packaged, outdated zls binary.

**Reply by williamboman** [reply]
Make sure zls is uninstalled in Mason

**Comment by rseichter** [comment]
Hm. If I uninstall Mason's zls plugin, LazyVim does not seem to call zls until I save the buffer. With the plugin active, I see code analysis while I am typing. You can probably tell that I am very new to LazyVim and its plugins, and I might therefore still have a general lack of understanding regarding the mechanics involved, so I ask for you patience. The following seems to work for me:
```lua
-- /.local/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/lsp/init.lua
"williamboman/mason.nvim",
cmd = "Mason",
keys = { { "<leader>cm", "<cmd>Mason<cr>", desc = "Mason" } },
opts = {
  -- Added by RS
  PATH = "append",
  -- [other options here ...]
},
```
I added `PATH = "append"` to the aforementioned init file, and this seems to ensure that my self-built `~/bin/zls` binary is used, because `~/bin` is the first entry in my shell's PATH variable.

**Reply by williamboman** [reply]
> LazyVim does not seem to call zls until I save the buffer

I'm afraid I don't know how LazyVim works and can't help further with that. When you set up the server via lspconfig it should register autocommands for the zig filetype. Might be a lazy loading issue, where servers are set up after the buffer is loaded?

> I added `PATH = "append"` to the aforementioned init file, and this seems to ensure that my self-built `~/bin/zls` binary is used, because `~/bin` is the first entry in my shell's PATH variable.

If you've uninstalled zls from Mason, there should be no `zls` in the Mason's bin directory, so setting it to `append` shouldn't be needed (although it makes sense to configure it this way if you'd like Neovim to prioritize "system" executables over Mason-installed ones)

**Reply by rseichter** [reply]
Using `PATH = "append"` allows me to keep the Mason zls plugin installed and working as intended, but in combination with the up-to-date `zls` binary I have built. That covers my primary requirement for now, although I would still prefer having a configuration option to specify an absolute path for Mason to use. Not knowing that the Mason plugin bundles a zls binary, let alone an outdated one, made me waste quite some time and also caused me to create an erroneous bug report in the ZLS project. That is not meant as a complaint, just to explain my asking for a `command = "/path/to/zls"` style configuration option for mason.nvim.

**Reply by williamboman** [reply]
Mason doesn't use any of the executables, it only installs them and makes them available in Neovim's `PATH`. If you've installed these elsewhere then there's no need to use Mason at all. The `command = "/path/to/zls"` is what you'd provide to the LSP client configuration (for example nvim-lspconfig or `:h vim.lsp.start()`).

**Reply by rseichter** [reply]
This just goes to show that I indeed don't know enough about the inner workings of the components involved yet. Perhaps I should not have switched from ye olde Vim to LazyVim and its collection of pre-configured plugins, but have gone the route of manually adding selected plugins to Neovim instead. I shall educate myself, and I thank you for your answers and patience.
